edb/server/protocol/auth_ext/oauth.py

```diff
@@ -0,0 +1,155 @@
+# limitations under the License.
+#
+
+
+import http
+import datetime
+import base64
+import immutables
+
+from typing import Type, TypeVar, overload
+from jwcrypto import jwt, jwk
+
+from edb import errors
+from edb.server.config import ops
+
+from . import github
+
+
+# TODO get this from edb.server.config.ops directly?
+SettingsMap = immutables.Map[str, ops.SettingValue]
+
+
+def redirect_to_auth_provider(
+    response,
+    redirect_uri: str,
+    iss: str,
+    provider_name: str,
+    db_config: SettingsMap,
+):
+    client_id, client_secret = _get_client_credentials(provider_name, db_config)
+    provider = _get_provider(
+        name=provider_name,
+        client_id=client_id,
+        client_secret=client_secret,
+    )
+
+    signing_key = _get_auth_signing_key(db_config)
+    state_token = _make_signed_token(
+        iss=iss, provider=provider_name, key=signing_key
+    )
+
+    auth_url = provider.get_code_url(
+        state=state_token.serialize(), redirect_uri=redirect_uri
+    )
+
+    response.status = http.HTTPStatus.FOUND
+    response.custom_headers["Location"] = auth_url
+    response.close_connection = True
+
+
+async def handle_auth_callback(request, response):
+    return
+
+
+providers = {
+    "github": github.GitHubProvider,
+}
+
+
+def _get_provider(name: str, client_id: str, client_secret: str):
+    provider_class = providers.get(name)
+    if provider_class is None:
+        raise errors.BackendError(f"Unknown provider: {name}")
+    return provider_class(client_id, client_secret)
+
+
+def _get_auth_signing_key(db_config: SettingsMap):
+    auth_signing_key = _maybe_get_config(db_config, "xxx_auth_signing_key")
+    if auth_signing_key is None:
+        raise errors.BackendError(
+            "No auth signing key configured: "
+            "Please set `cfg::Config.xxx_auth_signing_key`"
+        )
+    key_bytes = base64.b64encode(auth_signing_key.encode())
+
+    return jwk.JWK(kty="oct", k=key_bytes.decode())
+
+
+def _get_client_credentials(
+    provider_name: str, db_config: SettingsMap
+) -> tuple[str, str]:
+    client_id = _maybe_get_config(db_config, f"xxx_{provider_name}_client_id")
+    client_secret = _maybe_get_config(
+        db_config, f"xxx_{provider_name}_client_secret"
+    )
+    if client_id is None or client_secret is None:
+        raise errors.BackendError(
+            f"No client credentials configured for provider `{provider_name}`: "
+            f"Please set `cfg::Config.{provider_name}_client_id` and "
+            f"`cfg::Config.{provider_name}_client_secret`"
+        )
+    return (client_id, client_secret)
+
+
+T = TypeVar("T")
+
+
+@overload
+def _maybe_get_config(
+    db_config: SettingsMap, key: str, expected_type: Type[T]
+) -> T | None:
+    ...
+
+
+@overload
+def _maybe_get_config(db_config: SettingsMap, key: str) -> str | None:
+    ...
+
+
+def _maybe_get_config(
+    db_config: SettingsMap, key: str, expected_type: Type[object] = str
+) -> object:
```

Comment:
Now that you defined a `base.py`, I think this method will not be a part of the `oauth.py` implementation - other providers would use config too.

Reply:
Move this into `base.py`? Seems reasonable. I was thinking this would just hold the base class for just the provider, but if it's common practice to treat a module like this as a kind of utility module that holds some base functions and classes for the rest of the `auth_ext` module, that works for me!

Reply:
It's just a personal preference. It's just natural for me to think about the inheritance of `base` <- `oauth` <- `github`, where `base` has all the config stuff, `oauth` has the OAuth request handlers and `github` has only GitHub-specific stuff.

 It also makes total sense in Python to have flat function modules.

Reply:
Going to save this as a TODO for when refactoring of the OAuth and Provider classes.